### PR TITLE
Continuing nn p2

### PR DESCRIPTION
**Pull request by mqharris**

### CODE REVIEW

**Comment by mqharris** [comment]
Cleaned up and re-ran all the cells, which includes the model, so our f1, precision, recall changed very slightly

Some code, like the f1 score at the end of each epoch, is poorly optimized

This runs way faster for me in colab, so I'm thinking I'll leave the model notebook set up to run via colab

File: model/anomaly_nn.ipynb [hunk 1]
```diff
@@ -1,991 +1,1322 @@
...
+      "execution_count": 9,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "Ol2-sxy5-yUy"
+      },
+      "source": [
+        "# Convert to tensort and pad matrixes to have dims of 20x48 instead of 19x48\n",
```
Comment (WraySmith) [inline model/anomaly_nn.ipynb]: Is this still true for our dimensions? Should the padding be removed here - note that if it is removed we may need to modify some dimensions of the model (can always figure this out using print statements)

File: model/anomaly_nn.ipynb [hunk 2]
```diff
@@ -1,991 +1,1322 @@
...
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "ShW0RdNY-yU0"
+      },
+      "source": [
+        "# Hyperparameters\n",
+        "RANDOM_SEED = 1\n",
+        "# THESE ARE ALL STARTING POINTS BASED ON LAB3\n",
```
Comment (WraySmith) [inline model/anomaly_nn.ipynb]: remove comment - don't want references to labs

File: model/anomaly_nn.ipynb [hunk 3]
```diff
@@ -1,991 +1,1322 @@
...
+        "LEARNING_RATE = 0.001\n",
+        "BATCH_SIZE = 128\n",
+        "NUM_EPOCHS = 10\n",
+        "\n",
+        "# Architecture\n",
+        "# NUM_FEATURES = data_dims[0] * data_dims[1] # not sure this is needed currently\n",
+        "\n",
+        "# not sure we should be using an architecture with 2 classes with softmax\n",
+        "# doesn't it make more sense to have a single output node\n",
+        "# previous labs/lectures have 2 classes with softmax which is somewhat confusing\n",
```
Comment (WraySmith) [inline model/anomaly_nn.ipynb]: remove unneeded comments above - only leave comments that explain what was done as necessary

File: model/anomaly_nn.ipynb [hunk 4]
```diff
@@ -1,991 +1,1322 @@
...
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "KFxoCAHg-yU2",
+        "outputId": "130bb71b-b963-4527-dfec-bc3b8514ed9d"
+      },
+      "source": [
+        "device = torch.device(DEVICE)\n",
+        "torch.manual_seed(0)\n",
+        "\n",
+        "for epoch in range(2):\n",
```
Comment (WraySmith) [inline model/anomaly_nn.ipynb:412]: this code block needs comments - that it's just checking the train_loader (as discussed with Debangsha it doesn't have anything to do with warmup

File: model/anomaly_nn.ipynb [hunk 5]
```diff
@@ -1,991 +1,1322 @@
...
+        "    model.train()\n",
+        "    for batch_idx, (features, targets) in enumerate(train_loader):\n",
+        "        \n",
+        "        features = features.to(DEVICE)\n",
+        "        targets = targets.to(DEVICE, dtype=torch.long) # another had to use torch.long\n",
+        "            \n",
+        "        ### FORWARD AND BACK PROP\n",
+        "        logits, probas = model(features)\n",
+        "        \n",
+        "        cost = F.cross_entropy(logits, targets) # NEED TO MODIFY BASED ON BEST COST FUNCTION\n",
```
Comment (WraySmith) [inline model/anomaly_nn.ipynb]: I thought we were using binary_cross_entropy? Also need to remove comment

File: model/anomaly_nn.ipynb [hunk 6]
```diff
@@ -1,991 +1,1322 @@
...
+        "\n",
+        "        optimizer.zero_grad()\n",
+        "        \n",
+        "        cost.backward()\n",
+        "        minibatch_cost.append(cost) # from lecture 3 example files\n",
+        "        \n",
+        "        ### UPDATE MODEL PARAMETERS\n",
+        "        optimizer.step()\n",
+        "        \n",
+        "        ### LOGGING - MAY WANT TO MODIFY LOGGING INTERVALS\n",
```
Comment (WraySmith) [inline model/anomaly_nn.ipynb]: update comment

File: model/anomaly_nn.ipynb [hunk 7]
```diff
@@ -1,991 +1,1322 @@
...
+        "            print ('Epoch: %03d/%03d | Batch %04d/%04d | Cost: %.4f' \n",
+        "                   %(epoch+1, NUM_EPOCHS, batch_idx, \n",
+        "                     len(train_loader), cost, ))\n",
+        "\n",
+        "        \n",
+        "    model.eval()\n",
+        "    with torch.set_grad_enabled(False): # save memory during inference\n",
+        "        train_performance = compute_f1(model, train_loader, device=DEVICE)\n",
+        "        val_performance = compute_f1(model, val_loader, device=DEVICE)\n",
+        "        epoch_train_performance.append(train_performance) # from lecture 3 example files\n",
```
Comment (WraySmith) [inline model/anomaly_nn.ipynb]: remove comment

**Review by WraySmith — COMMENTED** [review]
Looks really good, just a couple of questions re: padding and the cost function used. Note that instead of removing padding, it may make sense to add 1 pad around the entire matrix.

Should also add in seeding anywhere there's randomness (shuffling etc) to ensure that the results don't change.

I also noticed that the data has versioning on it which should be removed. It may also be an idea to have an if/else statement re: running off of Colab vs locally (allowing a user to input `local  = True` or something like that to run it locally.
